src/hb-ot-shape-complex-machine-index.hh

```diff
@@ -0,0 +1,75 @@
+/*
+ *  This is part of HarfBuzz, a text shaping library.
+ *
+ * Permission is hereby granted, without written agreement and without
+ * license or royalty fees, to use, copy, modify, and distribute this
+ * software and its documentation for any purpose, provided that the
+ * above copyright notice and the following two paragraphs appear in
+ * all copies of this software.
+ *
+ * IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE TO ANY PARTY FOR
+ * DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
+ * ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN
+ * IF THE COPYRIGHT HOLDER HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * THE COPYRIGHT HOLDER SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING,
+ * BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
+ * ON AN "AS IS" BASIS, AND THE COPYRIGHT HOLDER HAS NO OBLIGATION TO
+ * PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+ */
+
+#ifndef HB_OT_SHAPE_COMPLEX_MACHINE_INDEX_HH
+#define HB_OT_SHAPE_COMPLEX_MACHINE_INDEX_HH
+
+#include "hb.hh"
+#include "hb-ot-shape-complex-use.hh"
+
+
+struct machine_index_t
+{
+  private:
+  hb_buffer_t *buffer;
+  unsigned int p;
+
+  public:
+  constexpr machine_index_t (hb_buffer_t *const buffer)
+    : buffer (buffer), p (0u) {}
+
+  unsigned int operator + (const int i) const
+  {
+    return p + i;
+  }
+
+  void operator = (const unsigned int i)
+  {
+    p = i;
+  }
+
+  bool operator == (const unsigned int i) const
+  {
+    return p == i;
+  }
+
+  unsigned int operator ++ ()
+  {
+    do {
+      ++p;
+    } while (p < buffer->len
+	     && (buffer->info[p].use_category() == USE_O || buffer->info[p].use_category() == USE_Rsv)
+	     && _hb_glyph_info_is_default_ignorable (&buffer->info[p]));
+    return p;
```

Comment:
Please make this class a template that takes a function-object to call to skip things, so we can use it from other shapers as well.

Reply:
Might even use it for AAT state machines...  Although I know @jfkthame had concerns there.